File: pip/skmap/src/skmap/head.py
#            ╓──────────┬──────────┬──────────┬──────────┐
#            ║  Byte 0  │  Byte 1  │  Byte 2  │  Byte 3  │
# ╒══════════╬══════════╧══════════╧══════════╧══════════╡
# │  Word 0  ║                                           │
# ├──────────╢                    ID          ┌──────────┤
# │  Word 1  ║                                │   Sync   │
# ├──────────╫──────────┬──────────┬──────────┴──────────┤
# │  Word 2  ║ Version  │  Flags   │       Checksum      │
# ├──────────╫──────────┼──────────┼──────────┬──────────┤
# │  Word 3  ║ Len_Kids │ Len_Sub  │  Len_K   │  Len_Var │
# └──────────╨──────────┴──────────┴──────────┴──────────┘

from dataclasses import dataclass

SIZE_WORD       : int = 4
SIZE_HEAD_WORDS : int = 4
SIZE_HEAD       : int = SIZE_HEAD_WORDS * SIZE_WORD
SIZE_ID         : int = 7
SIZE_CHECKSUM   : int = 2
SYNC            : int = 0xD8


@dataclass
class Head:

    id        : str
    sync      : int
    version   : int
    flags     : int
    checksum  : int
    len_kids  : int
    len_sub   : int
    len_k     : int
    len_var   : int


    def __init__(self, data : bytes):
        self.id        = data[0:SIZE_ID].decode()
        ii = SIZE_ID;
        self.sync      = data[ii]; ii += 1
        self.version   = data[ii]; ii += 1
        self.flags     = data[ii]; ii += 1
        self.checksum  = int.from_bytes(data[ii:ii+SIZE_CHECKSUM], byteorder='little')
        ii += SIZE_CHECKSUM
        self.len_kids  = data[ii]; ii += 1
        self.len_sub   = data[ii]; ii += 1
        self.len_k     = data[ii]; ii += 1
        self.len_var   = data[ii]; ii += 1
        # print(f'{data=}, {self.len_var=}')

    def valid_sync(self):
        return self.sync == SYNC

    def valid(self):
        return self.valid_sync()

    def __repr__(self) -> str:
        if not self.valid_sync():
            return f"{{BAD id={self.id.encode()}, {self.sync}}}"
        return f"{{{self.id}, v{self.version}, f={self.flags}, c={self.checksum_str()} l={self.len_sub}s+{self.len_kids}c+{self.len_k}k+{self.len_var}v}}"

    def checksum_str(self) -> str:
        return f'{self.checksum:0{SIZE_CHECKSUM}X}'

    def module_size_words(self) -> int:
        return SIZE_HEAD_WORDS + self.len_sub + self.len_kids + self.len_k + self.len_var

    def module_size(self) -> int:
        return self.module_size_words() * SIZE_WORD

File: sw/lib/cpp/skmap/src/head.cpp
#include "hdlskel/skmap/head.hpp"
#include <cassert>
#include <cstring>
#include <sstream>
#include <iomanip>

namespace {

std::ostream & to_hex(std::ostream & os, size_t w, uint val) {
    return os << std::hex << std::setw(w) << std::setfill('0') << val;
}

}

namespace hdlskel::skmap {

const Head unpack_head(std::span<const std::byte> head_bytes) {
    assert (head_bytes.size_bytes() == head_size);
    Head head;
    std::memcpy(&head, head_bytes.data(), head_size);
    return head;
}

std::string Head::id_str() const {
    return std::string(id.begin(), id.end());
}
std::string Head::checksum_str() const {
    std::ostringstream oss;
    to_hex(oss, 4, checksum);
    return oss.str();
}
// std::string id_str() const {
// }

std::string Head::to_str() const {
    std::ostringstream oss;

    if (not valid_sync()) {
        oss << "{BAD id=";

        // encode() → show raw bytes of id
        oss << "b'";
        for (unsigned char c : id) {
            // printable ASCII stays as-is, others hex-escaped
            if (std::isprint(c)) {
                oss << c;
            } else {
                oss << "\\x";
                to_hex(oss, 2, c);
            }
        }
        oss << "', ";
        to_hex(oss, 2, sync) << "}";
        return oss.str();
    }

    oss << "{" << id_str() << ", v" << static_cast<int>(version) << ", f=" << static_cast<int>(flags)
        << ", c=" << checksum_str()
        << " l=" << len_sub << "s+" << +len_kids << "c+" << +len_k << "k+" << +len_var << "v" << "}";

    return oss.str();
}

}
